### PR TITLE
implement namecheap dns management

### PR DESCRIPTION
**Pull request by rbelnap**
I was disappointed to see that the namecheap provider was only for registration, so I figured I'd take a stab at implementing it.  I've only known go for 2 days, and this is shamelessly cobbled together from other providers and bits of chewing gum.  It seems to do the basics, so I'm hopeful some more experienced contributors can work out the kinks before my mistakes get too ingrained.

### CODE REVIEW

**Comment by captncraig** [comment]
Awesome! I'll take a look at this shortly. Have you tried wiring it up to the provider integration test suite? That is a good way to make sure it covers a bunch of the corner cases we have found. Hopefully the readme in there is sufficient, but if not I can help.

File: providers/namecheap/namecheap.go [hunk 1]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+}
+
+func newReg(conf map[string]string) (providers.Registrar, error) {
+	return newProvider(conf, nil)
+}
+
+func splitDomain(domain string) (string, string) {
+	ds := strings.Split(domain, ".")
+	dl := len(ds)
+	tld := ds[dl-1]
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: This method fails for tlds like `.co.uk'. Should use https://godoc.org/golang.org/x/net/publicsuffix instead.

Its kinda crappy that their api expects SLD/TLD, and https://www.namecheap.com/support/api/methods/domains/get-info.aspx doesn't even split it like that for you.

File: providers/namecheap/namecheap.go [hunk 2]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
-func newReg(m map[string]string) (providers.Registrar, error) {
+func newDsp(conf map[string]string, metadata json.RawMessage) (providers.DNSServiceProvider, error) {
+	return newProvider(conf, metadata)
+}
+
+func newReg(conf map[string]string) (providers.Registrar, error) {
+	return newProvider(conf, nil)
+}
+
+func splitDomain(domain string) (string, string) {
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: I'd name the return values sld and tld for clarity. May also make sense to return an error in case we can't figure it out from the given domain.

File: providers/namecheap/namecheap.go [hunk 3]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+		return nil, err
+	}
+
+	var actual []*models.RecordConfig
+	for _, r := range records.Hosts {
+		if r.Type == "SOA" || r.Type == "NS" {
+			continue
+		}
+
+		if r.Name == "@" {
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: Shouldn't need to do this. Set Name to Name, and for NameFQDN use `dnsutil.AddOrigin` should be an example of that somewhere.

File: providers/namecheap/namecheap.go [hunk 4]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+
+	sld, tld := splitDomain(dc.Name)
+	records, err := n.client.DomainsDNSGetHosts(sld, tld)
+	if err != nil {
+		return nil, err
+	}
+
+	var actual []*models.RecordConfig
+	for _, r := range records.Hosts {
+		if r.Type == "SOA" || r.Type == "NS" {
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: Does namecheap not let you change NS records? Even for subdomains?

File: providers/namecheap/namecheap.go [hunk 5]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+
+	differ := diff.New(dc)
+	_, create, delete, modify := differ.IncrementalDiff(actual)
+
+	// because namecheap doesn't have selective create, delete, modify,
+	// we bundle them all up to send at once.  We *do* want to see the
+	// changes though
+
+	changes := false
+	desc := ""
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: Probably make desc a `[]string`, append each description to it, and then use `strings.Join(desc, "\n") to build a final string.

Then you can avoid the `changes` cariable altogether and use `if len(desc) > 0`

File: providers/namecheap/namecheap.go [hunk 6]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+					return n.UpdateRecords(dc)
+				},
+			})
+	}
+
+	return corrections, nil
+}
+
+func (n *Namecheap) GetNameservers(domainName string) ([]*models.Nameserver, error) {
+	// not exactly sure what is wanted here, but currently not needed (TODO)
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: This should return the nameservers to register if you are using namecheap's dns. Looks like maybe 

`"dns1.registrar-servers.com", "dns2.registrar-servers.com"` for the basic dns servers?

File: providers/namecheap/namecheap.go [hunk 7]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+// UpdateRecords bundles up the expected zone and sends it wholesale to
+// namecheap.
+func (n *Namecheap) UpdateRecords(dc *models.DomainConfig) error {
+
+	var recs []nc.DomainDNSHost
+
+	id := 1
+	for _, r := range dc.Records {
+		name := strings.Split(r.NameFQDN, "."+dc.Name)[0]
+		if name == dc.Name {
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: If you leave the Name as I recommend above, this may not need to be done either.

File: providers/namecheap/namecheap.go [hunk 8]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+			MXPref:  int(r.Priority),
+			TTL:     int(r.TTL),
+		}
+		recs = append(recs, rec)
+		id++
+	}
+
+	sld, tld := splitDomain(dc.Name)
+	result, err := n.client.DomainDNSSetHosts(sld, tld, recs)
+	fmt.Println("Namecheap returns:", result)
```
Comment (captncraig) [inline providers/namecheap/namecheap.go]: Remove debug before finalizing.

File: providers/namecheap/namecheap.go [hunk 9]
```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
...
+
+// UpdateRecords bundles up the expected zone and sends it wholesale to
+// namecheap.
+func (n *Namecheap) UpdateRecords(dc *models.DomainConfig) error {
+
+	var recs []nc.DomainDNSHost
+
+	id := 1
+	for _, r := range dc.Records {
+		name := strings.Split(r.NameFQDN, "."+dc.Name)[0]
```
Comment (tlimoncelli) [inline providers/namecheap/namecheap.go]: Are you trying to extract the name from a FQDN? (i.e. remove the domain suffix?).  I recommend using dnsutil.TrimDomainName() instead (warning: it FQDN == domain this returns "@" instead of "").

File: providers/namecheap/namecheap.go (hunk 6 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: I had suspected that and tried it, but namecheap doesn't like it.  When set that way : 
```CREATE NS dnscontrol.com dns1.registrar-servers.com. 300
CREATE NS dnscontrol.com dns2.registrar-servers.com. 300
RECREATING ZONE: dnscontrol.com
Namecheap returns: <nil>
FAILURE! Error 5050900: INVALID_NAME: NS record can't be set up for a @ or * host names
```
The best I could figure was to have it be a noop, but I'm open to interpretation

Comment (captncraig) [inline providers/namecheap/namecheap.go]: Yeah, this is pretty common. Too many providers control those way too tightly. The general policy in this case is to remove any NS records with Name `@` at the beginning of `GetDomainCorrections`. Something like:

```
dc.Filter(func(r *models.RecordConfig) bool {
    if r.Type == "NS" && r.Name == "@" {
        //check to make sure it looks like a namecheap server and print a warning if it does not.
        return false
    }
    return true
})
```

File: providers/namecheap/namecheap.go (hunk 5 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: Yeah, I can see that - I completely stole this section from the gandi provider, so the same can apply there ;)

File: providers/namecheap/namecheap.go (hunk 6 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: That would be a way to protect against that error in general, but I don't think it is applicable to the GetNameservers method, which it sounds like will need to remain a noop.

File: providers/namecheap/namecheap.go (hunk 9 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: Yes indeed, thanks for the tip! (fixed)

File: providers/namecheap/namecheap.go (hunk 2 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: fixed

File: providers/namecheap/namecheap.go (hunk 1 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: Agree on the crappy - implemented pubblicsuffix (hopefully correctly)

File: providers/namecheap/namecheap.go (hunk 4 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: It does! (for subdomains) fixed

File: providers/namecheap/namecheap.go (hunk 3 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: fixed

File: providers/namecheap/namecheap.go (hunk 6 above)
Comment (captncraig) [inline providers/namecheap/namecheap.go]: No, GetNameservers needs to be implemented. It does more than just create NS records. It also feeds the nameservers we give to the registrar part.

**Comment by captncraig** [comment]
Build is failing because the publicsuffix package needs to be copied into the repo. I just added it to master, so next time you push it should pass.

File: providers/namecheap/namecheap.go (hunk 6 above)
Comment (rbelnap) [inline providers/namecheap/namecheap.go]: Hmm.. so are you saying implement the defaults for GetNameservers, then drop it (or filter it out) in GetDomainCorrections?

Comment (captncraig) [inline providers/namecheap/namecheap.go]: Correct.

**Comment by rbelnap** [comment]
I think everything has been fixed here - I welcome further feedback.  I did wire it up to the integration tests, but it looks like namecheap adds in default records when a zone is emptied, and this breaks all tests.  Also note that I uncovered what I expect is a bug in the go-namecheap module and submitted it as https://github.com/billputer/go-namecheap/pull/6

**Comment by captncraig** [comment]
That sounds crappy. I'll try it when I get back to my desk and see what I can figure out. If that PR doesn't get merged quickly, it is possible to vendor in your changes regardless.

**Comment by captncraig** [comment]
Ok, I think I got past that particular issue, just by filtering out those records and pretending like they don't exist. 

Now I'm getting rate limited running the test suite.

Can you give me permission to push to your pr branch? It may be easier to just make a few edits directly.

**Comment by captncraig** [comment]
Oh, somehow I did push to that. Hmm.

**Comment by captncraig** [comment]
What the hell did I do? I can fix.

**Comment by captncraig** [comment]
I have no idea what I did. I was trying to add changes to your branch, but it did not do what I expected. 

If you re-push your changes to your repo, this may reopen. I'm sorry.

I also pushed your changes as the namecheap branch to this repo.

**Comment by captncraig** [comment]
I reopened it as #96. Its quite nearly ready anyway.
